File: VB/GoogleCalendarExample/CalendarImporter.vb
Imports Microsoft.VisualBasic
Imports Google.Apis.Auth.OAuth2
Imports Google.Apis.Calendar.v3
Imports Google.Apis.Calendar.v3.Data
Imports Google.Apis.Services
Imports Google.Apis.Util.Store
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Threading
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports DevExpress.XtraScheduler

Namespace GoogleCalendarExample
	Friend Class CalendarImporter
		Private storage_Renamed As SchedulerStorage
		Public Sub New(ByVal storage As SchedulerStorage)
			Me.storage_Renamed = storage
		End Sub

		Public ReadOnly Property Storage() As SchedulerStorage
			Get
				Return storage_Renamed
			End Get
		End Property

        Public Sub Import(ByVal events As IList(Of Google.Apis.Calendar.v3.Data.Event))
            Dim patternHash As New Dictionary(Of String, Appointment)()
            Dim occurrenceHash As New Dictionary(Of String, Google.Apis.Calendar.v3.Data.Event)()

            Storage.Appointments.BeginUpdate()
            Dim parser As New RecurrencePatternParser(storage_Renamed)
            For Each item As [Event] In events
                Dim appointment As Appointment = Nothing
                If item.RecurringEventId IsNot Nothing Then 'occurrence?
                    occurrenceHash.Add(item.Id, item) 'recurrence
                ElseIf item.Recurrence IsNot Nothing Then
                    appointment = parser.Parse(item.Recurrence, ConvertDateTime(item.Start), ConvertDateTime(item.End)) 'parse and create pattern
                    patternHash.Add(item.Id, appointment) 'normal appointment
                Else
                    appointment = storage_Renamed.CreateAppointment(AppointmentType.Normal)
                    AssingTimeIntervalPropertiesTo(appointment, item)
                End If
                If appointment Is Nothing Then
                    Continue For
                End If
                AssignCommonPropertiesTo(appointment, item)
                Storage.Appointments.Add(appointment)
            Next item
            LinkOccurrencesWithPatterns(occurrenceHash, patternHash)
            Storage.Appointments.EndUpdate()
        End Sub
        Private Sub LinkOccurrencesWithPatterns(ByVal occurrenceHash As Dictionary(Of String, Google.Apis.Calendar.v3.Data.Event), ByVal patternHash As Dictionary(Of String, Appointment))
            For Each occurrenceEntry As KeyValuePair(Of String, Google.Apis.Calendar.v3.Data.Event) In occurrenceHash
                Dim occurrenceEvent As [Event] = occurrenceEntry.Value
                Dim patternId As String = occurrenceEntry.Value.RecurringEventId
                If occurrenceEvent IsNot Nothing AndAlso patternHash.ContainsKey(patternId) Then
                    Dim pattern As Appointment = patternHash(patternId)
                    Dim exceptionType As AppointmentType = If((occurrenceEvent.Sequence Is Nothing), AppointmentType.DeletedOccurrence, AppointmentType.ChangedOccurrence)
                    If exceptionType.Equals(AppointmentType.DeletedOccurrence) Then
                        CreateDeletedOccurrence(pattern, occurrenceEvent)
                    Else
                        CreateChangedOccurrence(pattern, occurrenceEvent)
                    End If
                End If
            Next occurrenceEntry
        End Sub
		Private Sub CreateChangedOccurrence(ByVal pattern As Appointment, ByVal occurrenceEvent As [Event])
			Dim indx As Integer = CalculateOccurrenceIndex(pattern, occurrenceEvent.OriginalStartTime)
			Dim newOccurrence As Appointment = pattern.CreateException(AppointmentType.ChangedOccurrence, indx)
			newOccurrence.Start = ConvertDateTime(occurrenceEvent.Start)
			newOccurrence.End = ConvertDateTime(occurrenceEvent.End)
			AssignCommonPropertiesTo(newOccurrence, occurrenceEvent)
		End Sub
		Private Sub CreateDeletedOccurrence(ByVal pattern As Appointment, ByVal occurrenceEvent As [Event])
			Dim indx As Integer = CalculateOccurrenceIndex(pattern, occurrenceEvent.OriginalStartTime)
			Dim exception As Appointment = pattern.CreateException(AppointmentType.DeletedOccurrence, indx)
		End Sub

		Private Function CalculateOccurrenceIndex(ByVal pattern As Appointment, ByVal originalStartTime As EventDateTime) As Integer
			Dim calculator As OccurrenceCalculator = OccurrenceCalculator.CreateInstance(pattern.RecurrenceInfo)
			Return calculator.FindOccurrenceIndex(ConvertDateTime(originalStartTime), pattern)
		End Function
		Private Sub AssignCommonPropertiesTo(ByVal target As Appointment, ByVal source As [Event])
			target.Subject = source.Summary
			target.Description = source.Description
		End Sub
		Private Sub AssingTimeIntervalPropertiesTo(ByVal target As Appointment, ByVal source As [Event])
			If (Not source.Start.DateTime.HasValue) Then
				target.AllDay = True
			End If
			target.Start = ConvertDateTime(source.Start)
			target.End = ConvertDateTime(source.End)
		End Sub
		Private Function ConvertDateTime(ByVal start As EventDateTime) As DateTime
			If start.DateTime.HasValue Then
				Return start.DateTime.Value
			End If
			Return DateTime.Parse(start.Date)
		End Function
	End Class
End Namespace

File: CS/GoogleCalendarExample/CalendarImporter.cs
using Google.Apis.Auth.OAuth2;
using Google.Apis.Calendar.v3;
using Google.Apis.Calendar.v3.Data;
using Google.Apis.Services;
using Google.Apis.Util.Store;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;
using DevExpress.XtraScheduler;

namespace GoogleCalendarExample {
    class CalendarImporter {
        SchedulerStorage storage;
        public CalendarImporter(SchedulerStorage storage) {
            this.storage = storage;
        }

        public SchedulerStorage Storage { get { return storage; } }

        public void Import(IList<Event> events) {
            Dictionary<string, Appointment> patternHash = new Dictionary<string, Appointment>();
            Dictionary<string, Event> occurrenceHash = new Dictionary<string, Event>();

            Storage.Appointments.BeginUpdate();
            RecurrencePatternParser parser = new RecurrencePatternParser(storage);
            foreach(Event item in events) {
                Appointment appointment = null;
                if(item.RecurringEventId != null) { //occurrence?
                    occurrenceHash.Add(item.Id, item);
                } else if(item.Recurrence != null) { //recurrence                    
                    appointment = parser.Parse(item.Recurrence, ConvertDateTime(item.Start), ConvertDateTime(item.End)); //parse and create pattern
                    patternHash.Add(item.Id, appointment);
                } else { //normal appointment
                    appointment = storage.CreateAppointment(AppointmentType.Normal);
                    AssingTimeIntervalPropertiesTo(appointment, item);
                }
                if(appointment == null)
                    continue;
                AssignCommonPropertiesTo(appointment, item);
                Storage.Appointments.Add(appointment);
            }
            LinkOccurrencesWithPatterns(occurrenceHash, patternHash);
            Storage.Appointments.EndUpdate();
        }
        void LinkOccurrencesWithPatterns(Dictionary<string, Event> occurrenceHash, Dictionary<string, Appointment> patternHash) {
            foreach(KeyValuePair<string, Event> occurrenceEntry in occurrenceHash) {
                Event occurrenceEvent = occurrenceEntry.Value;
                string patternId = occurrenceEntry.Value.RecurringEventId;
                if(occurrenceEvent != null && patternHash.ContainsKey(patternId)) {
                    Appointment pattern = patternHash[patternId];
                    AppointmentType exceptionType = (occurrenceEvent.Sequence == null) ? AppointmentType.DeletedOccurrence : AppointmentType.ChangedOccurrence;
                    if(exceptionType == AppointmentType.DeletedOccurrence)
                        CreateDeletedOccurrence(pattern, occurrenceEvent);
                    else
                        CreateChangedOccurrence(pattern, occurrenceEvent);
                }
            }
        }
        void CreateChangedOccurrence(Appointment pattern, Event occurrenceEvent) {
            int indx = CalculateOccurrenceIndex(pattern, occurrenceEvent.OriginalStartTime);
            Appointment newOccurrence = pattern.CreateException(AppointmentType.ChangedOccurrence, indx);
            newOccurrence.Start = ConvertDateTime(occurrenceEvent.Start);
            newOccurrence.End = ConvertDateTime(occurrenceEvent.End);
            AssignCommonPropertiesTo(newOccurrence, occurrenceEvent);
        }
        void CreateDeletedOccurrence(Appointment pattern, Event occurrenceEvent) {
            int indx = CalculateOccurrenceIndex(pattern, occurrenceEvent.OriginalStartTime);
            Appointment exception = pattern.CreateException(AppointmentType.DeletedOccurrence, indx);
        }

        int CalculateOccurrenceIndex(Appointment pattern, EventDateTime originalStartTime) {
            OccurrenceCalculator calculator = OccurrenceCalculator.CreateInstance(pattern.RecurrenceInfo);
            return calculator.FindOccurrenceIndex(ConvertDateTime(originalStartTime), pattern);
        }
        void AssignCommonPropertiesTo(Appointment target, Event source) {
            target.Subject = source.Summary;
            target.Description = source.Description;
        }
        void AssingTimeIntervalPropertiesTo(Appointment target, Event source) {
            if(!source.Start.DateTime.HasValue)
                target.AllDay = true;
            target.Start = ConvertDateTime(source.Start);
            target.End = ConvertDateTime(source.End);
        }
        DateTime ConvertDateTime(EventDateTime start) {
            if(start.DateTime.HasValue)
                return start.DateTime.Value;
            return DateTime.Parse(start.Date);
        }        
    }
}
